custom/icds_reports/utils/aggregation_helpers/distributed/agg_dashboard_acitivity.py

```diff
@@ -0,0 +1,226 @@
+from datetime import timedelta
+from django.db.models import Max
+
+import re
+
+from custom.icds_reports.utils.aggregation_helpers.distributed.base import BaseICDSAggregationDistributedHelper
+from custom.icds_reports.const import  AGG_DASHBOARD_ACTIVITY
+from corehq.apps.users.dbaccessors.all_commcare_users import get_all_user_id_username_pairs_by_domain
+from django.utils.functional import cached_property
+from corehq.apps.users.models import CommCareUser
+
+
+class DashboardActivityReportAggregate(BaseICDSAggregationDistributedHelper):
+    aggregate_parent_table = AGG_DASHBOARD_ACTIVITY
+
+    def __init__(self, date):
+        self.date = date
+        self.last_agg_date = self.get_last_agg_date()
+
+    def aggregate(self, cursor):
+        cursor.execute(self.drop_table_query())
+        cursor.execute(*self.create_table_query())
+        add_query, add_params = self.add_missing_users()
+        args_str = ','.join(cursor.mogrify("(%s,%s,%s,%s,%s,%s)", x).decode('utf-8') for x in add_params['values'])
+
+        cursor.execute(add_query + args_str)
+
+        cursor.execute(*self.rollover_previous_data())
+
+        update_queries = self.update_query()
+
+        for query, param in update_queries:
+            cursor.execute(query, param)
+
+    @cached_property
+    def get_dashboard_users(self):
+        all_users = get_all_user_id_username_pairs_by_domain(self.domain, include_web_users=False,
+                                                             include_mobile_users=True)
+        dashboard_uname_rx = re.compile(r'^\d*\.[a-zA-Z]*@.*')
+
+        return {
+            uname
+            for id, uname in all_users
+            if dashboard_uname_rx.match(uname)
+        }
+
+    def get_user_locations(self):
```

Comment:
This does one or more queries per user and will be quite slow now, and even slower in the future. You might be able to utilize something like `AwcLocation` to get all this info at once and then do the matching in python

Reply:
done